File: test-suite/python/security/header_injection_clean.py
import urllib.parse

from django.http import FileResponse, HttpResponse
from flask import Response, make_response, request, send_file
from werkzeug.utils import secure_filename


def sanitize_header_value(value):
    cleaned = str(value).replace("\r", "").replace("\n", "")
    if not cleaned:
        raise ValueError("empty header value")
    return cleaned


def flask_header_assignment_clean():
    response = make_response("ok")
    response.headers["X-Display-Name"] = sanitize_header_value(request.args["name"])
    return response


def django_header_assignment_clean(django_request):
    response = HttpResponse("ok")
    safe_filename = urllib.parse.quote(django_request.GET["filename"], safe="")
    response["Content-Disposition"] = f'attachment; filename="{safe_filename}"'
    return response


def response_headers_dict_clean():
    return Response("ok", headers={"X-Trace": "internal"})


def response_headers_variable_clean():
    safe_headers = {"X-Trace": urllib.parse.quote(request.args["trace"], safe="")}
    return Response("ok", headers=safe_headers)


def local_headers_subscript_clean():
    headers = {}
    headers["X-Trace"] = urllib.parse.quote(request.args["trace"], safe="")
    return Response("ok", headers=headers)


def headers_add_method_clean():
    response = make_response("ok")
    token = urllib.parse.quote(request.headers["X-Trace"], safe="")
    response.headers.add("X-Trace", token)
    return response


def headers_update_method_clean():
    response = make_response("ok")
    response.headers.update({"X-Session": sanitize_header_value(request.args["session"])})
    return response


def flask_download_name_clean():
    safe_download_name = secure_filename(request.args["name"])
    return send_file("/srv/reports/monthly.pdf", as_attachment=True, download_name=safe_download_name)


def django_file_response_clean():
    return FileResponse(iter([b"report"]), as_attachment=True, filename="monthly.pdf")

File: test-suite/js/security/header-injection-clean.ts
type RequestLike = {
  query: Record<string, string | undefined>;
  body: Record<string, string | undefined>;
  headers: Record<string, string | undefined>;
  nextUrl: { searchParams: URLSearchParams };
};

type ResponseLike = {
  setHeader(name: string, value: string): void;
  header(name: string, value: string): void;
};

function sanitizeHeaderValue(value: string | null | undefined): string {
  const clean = String(value ?? "").replace(/[\r\n]/g, "");
  if (clean.length === 0) {
    throw new Error("empty header value");
  }
  return clean;
}

function encodeFilename(filename: string | undefined): string {
  return encodeURIComponent(filename ?? "report.pdf");
}

export function expressHeader(req: RequestLike, res: ResponseLike): void {
  res.setHeader("X-Display-Name", sanitizeHeaderValue(req.query.name));
}

export function contentDisposition(req: RequestLike, res: ResponseLike): void {
  const filename = encodeFilename(req.query.filename);
  res.setHeader("Content-Disposition", `attachment; filename*=UTF-8''${filename}`);
}

export function fastifyReply(req: RequestLike, reply: ResponseLike): void {
  reply.header("X-Trace", sanitizeHeaderValue(req.headers["x-trace-id"]));
}

export function responseObject(request: RequestLike): Response {
  const token = request.nextUrl.searchParams.get("token");
  return new Response("ok", {
    headers: {
      "X-Token": sanitizeHeaderValue(token),
    },
  });
}

export function headersApi(req: RequestLike): Headers {
  const responseHeaders = new Headers();
  responseHeaders.set("X-Session", sanitizeHeaderValue(req.body.session));
  return responseHeaders;
}
